include/hax_interface.h

```diff
@@ -49,6 +49,124 @@
+#define HAX_VM_IOCTL_SET_RAM__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0x82, 0x904, struct hax_set_ram_info)
+#define HAX_VM_IOCTL_VCPU_DESTROY__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOR,  0x83, 0x905, uint32_t)
+#define HAX_VM_IOCTL_ADD_RAMBLOCK__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOW,  0x85, 0x913, struct hax_ramblock_info)
+#define HAX_VM_IOCTL_SET_RAM2__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0x86, 0x914, struct hax_set_ram_info2)
+#define HAX_VM_IOCTL_PROTECT_RAM__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0x87, 0x915, struct hax_protect_ram_info)
+
+#define HAX_VCPU_IOCTL_RUN__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IO,   0xc0, 0x906, HAX_UNUSED)
+#define HAX_VCPU_IOCTL_SETUP_TUNNEL__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0xc5, 0x90b, struct hax_tunnel_info)
+#define HAX_VCPU_IOCTL_GET_REGS__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0xc8, 0x90e, struct vcpu_state_t)
+#define HAX_VCPU_IOCTL_SET_REGS__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0xc7, 0x90d, struct vcpu_state_t)
+#define HAX_VCPU_IOCTL_GET_FPU__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOR,  0xc4, 0x90a, struct fx_layout)
+#define HAX_VCPU_IOCTL_SET_FPU__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOW,  0xc3, 0x909, struct fx_layout)
+#define HAX_VCPU_IOCTL_GET_MSRS__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0xc2, 0x908, struct hax_msr_data)
+#define HAX_VCPU_IOCTL_SET_MSRS__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0xc1, 0x907, struct hax_msr_data)
+#define HAX_VCPU_IOCTL_INTERRUPT__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOWR, 0xc6, 0x90c, uint32_t)
+
+// API 2.0
+#define HAX_VM_IOCTL_NOTIFY_QEMU_VERSION__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOW,  0x84, 0x910, struct hax_qemu_version)
+#define HAX_VCPU_IOCTL_DEBUG__LEGACY \
+    HAX_LEGACY_IOCTL(HAX_IOW,  0xc9, 0x916, struct hax_debug_t)
+
+/* API
+ * ===
+ * Each platform generates their own IOCTL-value by using the macro
+ * HAX_IOCTL(access, code, type) with the following arguments:
+ * - access: Arguments usage from userland perspective.
+ *   - HAX_IO:    Driver ignores user arguments.
+ *   - HAX_IOR:   Driver writes user arguments (read by user).
+ *   - HAX_IOW:   Driver reads user arguments (written by user).
+ *   - HAX_IOWR:  Driver reads+writes user arguments (written+read by user).
+ * - code: Sequential number in range 0x00-0x3F, and maskable via:
+ *   - HAX_IOCTL_PLATFORM  (0x40)  Platform-specific ioctl.
+ *   - HAX_IOCTL_EXTENSION (0x80)  Extension-specific ioctl.
+ * - type: User argument type.
+ */
+#define HAX_IOCTL_VERSION \
+    HAX_IOCTL(HAX_IOWR, 0x00, struct hax_module_version)
+#define HAX_IOCTL_CREATE_VM \
+    HAX_IOCTL(HAX_IOWR, 0x01, uint32_t)
+#define HAX_IOCTL_DESTROY_VM \
+    HAX_IOCTL(HAX_IOW,  0x02, uint32_t)
+#define HAX_IOCTL_CAPABILITY \
+    HAX_IOCTL(HAX_IOR,  0x03, struct hax_capabilityinfo)
+#define HAX_IOCTL_SET_MEMLIMIT \
+    HAX_IOCTL(HAX_IOWR, 0x04, struct hax_set_memlimit)
+
+#define HAX_VM_IOCTL_VCPU_CREATE \
+    HAX_IOCTL(HAX_IOWR, 0x00, uint32_t)
+#define HAX_VM_IOCTL_ALLOC_RAM \
```

Comment:
This IOCTL has been replaced by `HAX_VM_IOCTL_ADD_RAMBLOCK`, and both QEMU and Android Emulator now use the latter. So I wouldn't include it in the new API.

Reply:
On the other hand, although `HAX_VM_IOCTL_SET_RAM2` is meant to replace `HAX_VM_IOCTL_SET_RAM`, QEMU is still using the latter, so please define both for the new interface.

Reply:
Sure, my upcoming QEMU patch will make sure `HAX_VM_IOCTL_SET_RAM2` is used. After the grace period we can rename `HAX_VM_IOCTL_SET_RAM2` to `HAX_VM_IOCTL_SET_RAM`. Since it's only a change of name (and not of the ioctl value or user argument type), it will happen seamlessly for users.

Reply:
Good idea, thanks!